src/main/scala/com/fulcrumgenomics/umi/CorrectUmis.scala

```diff
@@ -0,0 +1,229 @@
+    |
+    |All of the fixed UMIs must be of he same length, and all UMIs in the BAM file must also have the same
+    |length.  Multiple UMIs that are concatenated with hyphens (e.g. AACCAGT-AGGTAGA) are split apart,
+    |corrected individually and then re-assembled.  A read is accepted only if all the UMIs can be corrected.
+    |
+    |Correction is controlled by two parameters:
+    |  1. --max-mismatches controls how many mismatches (no-calls are counted as mismatches) are tolerated
+    |         between a UMI as read and a fixed UMI.
+    |  2. --min-distance controls how many more mismatches the next best hit must have
+    |
+    |For example, with two fixed UMIs AAAAA and CCCCC and max-mismatches=3 and min-distance=2 the
+    |following would happen:
+    |  - AAAAA would match to AAAAA
+    |  - AAGTG would match to AAAAA with three mismatches because CCCCCC has six mismatches and 6 >= 3 + 2
+    |  - AACCA would be rejected because it is 2 mismatches to AAAAA and 3 to CCCCCC and 3 <= 2 + 2
+    |
+    |The set of fixed UMIs may be specified on the command line using --umis umi1 umi2 ... or via one or
+    |more files of UMIs with a single sequence per line using --umi-files umis.txt more_umis.txt
+  """)
+class CorrectUmis
+( @arg(flag="i", doc="Input SAM or BAM file.")  val input: PathToBam,
+  @arg(flag="o", doc="Output SAM or BAM file.") val output: PathToBam,
+  @arg(flag="r", doc="Reject BAM file to save unassigned reads.") val rejects: Option[PathToBam] = None,
+  @arg(flag="M", doc="Metrics file to write.") val metrics: Option[FilePath] = None,
+  @arg(flag="m", doc="Maximum number of mismatches between a UMI and an expected UMI.") val maxMismatches: Int,
+  @arg(flag="d", doc="Minimum distance (in mismatches) to next best UMI.") val minDistance: Int,
+  @arg(flag="u", doc="Expected UMI sequences.", minElements=0) val umis: Seq[String] = Seq.empty,
+  @arg(flag="U", doc="File of UMI sequences, one per line.", minElements=0) val umiFiles: Seq[FilePath] = Seq.empty,
+  @arg(flag="t", doc="Tag in which UMIs are stored.") val umiTag: String = ConsensusTags.UmiBases
+) extends FgBioTool with LazyLogging {
+
+  validate(umis.nonEmpty || umiFiles.nonEmpty, "At least one UMI or UMI file must be provided.")
+  Io.assertReadable(input)
+  Io.assertReadable(umiFiles)
+  Io.assertCanWriteFile(output)
+  rejects.foreach(Io.assertCanWriteFile(_))
+
+  override def execute(): Unit = {
+    // Construct the full set of UMI sequences to match again
+    val (umiSequences, umiLength) = {
+      val set = mutable.HashSet[String](umis:_*)
+      umiFiles.foreach(Io.readLines(_).map(_.trim).filter(_.nonEmpty).foreach(set.add))
+      validate(set.nonEmpty, s"At least one UMI sequence must be provided; none found in files ${umiFiles.mkString(", ")}")
+
+      val lengths = set.map(_.length)
+      validate(lengths.size == 1, s"UMIs of multiple lengths found. Lengths: ${lengths.mkString(", ")}")
+      (set.toArray, lengths.head)
```

Comment:
I agree the time to think about it is during design, but I don't think we can always rely on the design being good since we cannot always perform the design ourselves.  Thinking about it more, I want like to see a double check:
```
set.toSeq
    .combinations(2)
    .map { vec => (vec(0), vec(1) }  // for readability
    .map { case (first, second) => 
        val numMismatches = Sequences.countMismatches(first, second)
        if (numMismatches < minDistance) {
            log.warn(s"Umis are within min-distance $minDistance: $first and $second have $numMismatches distance")
        }
    }
```